scripts/common/smoketest.go

```diff
@@ -0,0 +1,155 @@
+
+type SmokeTestOpts struct {
+	LogDir     string
+	RepoDir    string
+	NumWorkers int
+	Compile    func() error
+	Run        func() error
+}
+
+func (s *SmokeTest) InitOptions() (*SmokeTestOpts, error) {
+	logDir := flag.String("logdir", "/tmp/scoot-smoketest", "If empty, write to stdout, else to log file")
+	repoDir := flag.String("repo", "$GOPATH/src/github.com/scootdev/scoot", "Path to scoot repo")
+	numWorkers := flag.Int("num_workers", 5, "Number of workerserver instances to spin up.")
+
+	flag.Parse()
+	if *numWorkers < 5 {
+		return nil, errors.New("Need >5 workers (see scheduler.go:getNumNodes)")
+	}
+	compile := func() error { return s.compile(*repoDir) }
+	run := func() error { return s.run(*numWorkers) }
+	opts := &SmokeTestOpts{*logDir, *repoDir, *numWorkers, compile, run}
+	return opts, nil
+}
+
+func (s *SmokeTest) Fatalf(format string, v ...interface{}) {
+	fmt.Printf(format+"\n", v...)
+	s.Kill()
+}
+
+func (s *SmokeTest) Kill() {
+	for _, cmd := range *s {
+		if cmd.ProcessState == nil && cmd.Process != nil {
+			fmt.Printf("Killing: %v\n", cmd.Path)
+			p, _ := os.FindProcess(cmd.Process.Pid)
+			p.Signal(os.Interrupt)
+		}
+	}
+}
+
+func (s *SmokeTest) GetPort() int {
+	l, _ := net.Listen("tcp", "localhost:0")
+	defer l.Close()
+	return l.Addr().(*net.TCPAddr).Port
+}
+
+func (s *SmokeTest) RunCmd(blocking bool, name string, args ...string) error {
+	bin := filepath.Join(os.Getenv("GOPATH"), "bin", name)
+	cmd := exec.Command(bin, args...)
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+	*s = append(*s, cmd)
+	if err := cmd.Start(); err != nil {
+		return fmt.Errorf("Error starting %v,%v: %v", bin, args, err)
+	}
+	fmt.Println("Started: ", name, args)
+
+	if !blocking {
+		go func() {
+			if err := cmd.Wait(); err != nil {
+				s.Fatalf("Exiting %v,%v: %v", bin, args, err)
+			}
+		}()
+		return nil
+	} else {
+		return cmd.Wait()
+	}
+}
+
+func (s *SmokeTest) SetupLog(logDir string) error {
+	if logDir != "" {
+		if err := os.MkdirAll(logDir, 0777); err != nil {
+			return fmt.Errorf("Could not create log dir: %v", logDir)
+		}
+		logpath := path.Join(logDir, "smoketest"+strconv.Itoa(os.Getpid()))
+		out, err := os.Create(logpath)
+		if err != nil {
+			return fmt.Errorf("Could not init logfile: %v", logpath)
+		}
+		fmt.Println("Redirecting to log: ", logpath)
+		os.Stdout = out
+		os.Stderr = out
+	}
+	return nil
+}
+
+func (s *SmokeTest) StartSignalHandler() {
+	sigchan := make(chan os.Signal, 1)
+	signal.Notify(sigchan, syscall.SIGINT, syscall.SIGTERM)
+	go func() {
+		var sig os.Signal
+		sig = <-sigchan
+		fmt.Printf("signal %s received, shutting down", sig)
+		s.Kill()
+		os.Exit(1)
+	}()
+}
+
+func (s *SmokeTest) compile(repoDir string) error {
+	cmd := exec.Command("go", "install", "./binaries/...")
+	cmd.Dir = os.ExpandEnv(repoDir)
+	out, err := cmd.CombinedOutput()
+	if err != nil {
+		fmt.Println(string(out))
+		return err
+	}
+	return nil
+}
+
+func (s *SmokeTest) run(numWorkers int) error {
```

Comment:
Seems like we can continue using the name 'Run()' if I rename SmokeTest to SwarmTest?